### PR TITLE
Enable user to supply alternative trust store for Vault

### LINKED ISSUE #780: Support HTTPs vault instances and alternative sources of trust
Currently, we are testing integration with a HashiCorp Vault instance operating over plain HTTP.  Users will want to use HTTPs to meet their security requirements.

1. We should validate that HTTPs works (it probably will) and add integration tests.
1. Some users might require an alternative source of trust, so we should allow its configuration.  The integration tests might need the same.
1. We should investigate if HashCorp Vault or HCP Vault supports mTLS.   If so, we might chose to defer the feature.

### PR DESCRIPTION
**Pull request by robobario**
### Type of change

- Enhancement / new feature

### Description

Users can supply an alternative source of trust for Vault.

Resolves #780

To use insecure authentication (for development!):
```yaml
kmsConfig:
  tls:
    trust:
      insecure: true
```

To use platform default keystore type with inline password:
```yaml
kmsConfig:
  tls:
    trust:
      storeFile: /path/to/trust.jks
      storePasswordProvider:
        password: changeit
```

and with password from a file:
```yaml
kmsConfig:
  tls:
    trust:
      storeFile: /path/to/trust.jks
      storePasswordProvider:
        passwordFile: /path/to/password.file
```

and with custom store type:
```yaml
kmsConfig:
  tls:
    trust:
      storeFile: /path/to/trust.p12
      storePasswordProvider:
        password: changeit
      storeType: PKCS12
```

Why:
The vault server may be protected by a private certificate so we should support the user supplying an alternative trust store.

### Checklist

_Please go through this checklist and make sure all applicable tasks have been done_

- [ ] Write tests
- [ ] Make sure all tests pass
- [ ] Review performance test results. Ensure that any degradations to performance numbers are understood and justified.
- [ ] Make sure all Sonar-Lint warnings are addressed or are justifiably ignored.
- [ ] Update documentation
- [ ] Reference relevant issue(s) and close them after merging
- [ ] For user facing changes, update CHANGELOG.md (remember to include changes affecting the API of the test artefacts too).

### CODE REVIEW

File: kroxylicious-kms-provider-hashicorp-vault/src/main/java/io/kroxylicious/kms/provider/hashicorp/vault/config/InsecureTls.java [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+package io.kroxylicious.kms.provider.hashicorp.vault.config;
+
+import java.security.cert.X509Certificate;
+
+import javax.net.ssl.TrustManager;
+import javax.net.ssl.X509TrustManager;
+
+/**
+ * A {@link TrustProvider} that allows the disabling trust verification.
+ * <strong>Not recommended for production use.</strong>
```
Comment (SamBarker) [inline kroxylicious-kms-provider-hashicorp-vault/src/main/java/io/kroxylicious/kms/provider/hashicorp/vault/config/InsecureTls.java:16]: I think we should have a stronger warning. 
```suggestion
 * <strong>Should not be used outside of automated tests.</strong>
```

Is there any way we can shift if out of the production source tree?

File: kroxylicious-kms-provider-hashicorp-vault/src/main/java/io/kroxylicious/kms/provider/hashicorp/vault/config/TrustProvider.java [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+ */
+@JsonTypeInfo(use = JsonTypeInfo.Id.DEDUCTION)
+@JsonSubTypes({ @JsonSubTypes.Type(TrustStore.class), @JsonSubTypes.Type(InsecureTls.class) })
+public interface TrustProvider {
+
+    /**
+     * Return trust managers
+     * @return trust manager array, or null if this provider supplies no trust
+     */
+    TrustManager[] getTrustManagers();
```
Comment (SamBarker) [inline kroxylicious-kms-provider-hashicorp-vault/src/main/java/io/kroxylicious/kms/provider/hashicorp/vault/config/TrustProvider.java:25]: When do we want to support multiple `TrustManager`s? Can we use Optional<TrustManager[]>` instead of a semantic null?

File: kroxylicious-kms-provider-hashicorp-vault/src/test/java/io/kroxylicious/kms/provider/hashicorp/vault/VaultKmsIT.java [hunk 3]
```diff
@@ -44,35 +60,81 @@ class VaultKmsIT {
...
 
+    private static void startVault() {
         vaultContainer = new VaultContainer<>(HASHICORP_VAULT)
                 .withVaultToken(VAULT_TOKEN)
                 .withEnv("VAULT_FORMAT", "json")
+                // in vault server -dev mode the 8200 port cannot be reconfigured to TLS and is also useful so that the
+                // internal vault commands we make while execed into the container can communicate with vault over plain
+                // without having to trust the TLS cert within the container. Plus we can test TLS and non-TLS against the
+                // same container.
+                .withExposedPorts(PLAIN_PORT, TLS_PORT)
```
Comment (SamBarker) [inline kroxylicious-kms-provider-hashicorp-vault/src/test/java/io/kroxylicious/kms/provider/hashicorp/vault/VaultKmsIT.java:104]: The rational for enabling `-dev` mode makes sense but should we always expose both ports? I guess there is no harm in always exposing the `TLS_PORT` but do we risk false positives in tests if we leave the `PLAIN_PORT` exposed when TLS is enabled?

File: kroxylicious-kms-provider-hashicorp-vault/src/test/java/io/kroxylicious/kms/provider/hashicorp/vault/VaultKmsIT.java [hunk 4]
```diff
@@ -44,35 +60,81 @@ class VaultKmsIT {
...
+                .withCopyFileToContainer(MountableFile.forClasspathResource("Vault.hcl"), "/vault/config/Vault.hcl")
+                .withCopyFileToContainer(MountableFile.forHostPath(keys.serverCertPem(), 0777), "/vault/config/cert.pem")
+                .withCopyFileToContainer(MountableFile.forHostPath(keys.serverPrivateKeyPem(), 0777), "/vault/config/key.pem")
```
Comment (SamBarker) [inline kroxylicious-kms-provider-hashicorp-vault/src/test/java/io/kroxylicious/kms/provider/hashicorp/vault/VaultKmsIT.java:107]: I know, I know, its just a test container, but we really shouldn't need write permissions to certificate files!

**Review by SamBarker — COMMENTED** [review]
I'm a bit concerned about including `InsecureTls` in the production Vault provider otherwise I'm happy to merge this PR.

File: kroxylicious-kms-provider-hashicorp-vault/src/main/java/io/kroxylicious/kms/provider/hashicorp/vault/config/InsecureTls.java (hunk 1 above)
Comment (k-wall) [reply kroxylicious-kms-provider-hashicorp-vault/src/main/java/io/kroxylicious/kms/provider/hashicorp/vault/config/InsecureTls.java:16]: I wouldn't shift InsecureTls out from the tree.   

- it is feature people expect to find amongst TLS configuration options. 
- we've done our bit and documented the recommendation not to use in production.  I think the responsibility now rests with the production deployer to make good choices.
- i know I've valued have insecureTls options to aid problem diagnosis. it is a useful blade to have on the swiss army knife.

File: kroxylicious-kms-provider-hashicorp-vault/src/test/java/io/kroxylicious/kms/provider/hashicorp/vault/VaultKmsIT.java (hunk 4 above)
Comment (k-wall) [reply kroxylicious-kms-provider-hashicorp-vault/src/test/java/io/kroxylicious/kms/provider/hashicorp/vault/VaultKmsIT.java:107]: +1

yeah.. before you know it two things happen:  a) code scanning tools go red b) someone copies the code to an app that ends up in production.

File: kroxylicious-kms-provider-hashicorp-vault/src/main/java/io/kroxylicious/kms/provider/hashicorp/vault/config/Tls.java [hunk 5]
```diff
@@ -0,0 +1,59 @@
+/*
+ * Copyright Kroxylicious Authors.
+ *
+ * Licensed under the Apache Software License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+package io.kroxylicious.kms.provider.hashicorp.vault.config;
```
Comment (k-wall) [inline kroxylicious-kms-provider-hashicorp-vault/src/main/java/io/kroxylicious/kms/provider/hashicorp/vault/config/Tls.java:7]: I was imagining that the Tls class (and friends) would move to kroxylicious-api avoiding the need to duplicate the class.  It would mean that any filter and any kms implementation could re-use.

File: kroxylicious-kms-provider-hashicorp-vault/src/test/java/io/kroxylicious/kms/provider/hashicorp/vault/CertificateGenerator.java [hunk 6]
```diff
@@ -0,0 +1,176 @@
...
+
+import org.bouncycastle.asn1.x500.X500Name;
+import org.bouncycastle.asn1.x509.SubjectPublicKeyInfo;
+import org.bouncycastle.cert.X509CertificateHolder;
+import org.bouncycastle.cert.X509v3CertificateBuilder;
+import org.bouncycastle.cert.jcajce.JcaX509CertificateConverter;
+import org.bouncycastle.jce.provider.BouncyCastleProvider;
+import org.bouncycastle.openssl.jcajce.JcaPEMWriter;
+import org.bouncycastle.operator.jcajce.JcaContentSignerBuilder;
+import org.testcontainers.shaded.com.google.common.io.Files;
```
Comment (k-wall) [inline kroxylicious-kms-provider-hashicorp-vault/src/test/java/io/kroxylicious/kms/provider/hashicorp/vault/CertificateGenerator.java:35]: Don't use shaded

**Review by k-wall — APPROVED** [review]

File: kroxylicious-kms-provider-hashicorp-vault/src/main/java/io/kroxylicious/kms/provider/hashicorp/vault/config/TrustProvider.java (hunk 2 above)
Comment (k-wall) [reply kroxylicious-kms-provider-hashicorp-vault/src/main/java/io/kroxylicious/kms/provider/hashicorp/vault/config/TrustProvider.java:25]: The TMF returns an array. I'm not sure what circumstances may it can return >1.  Maybe sticking with an array is safer.
https://docs.oracle.com/en/java/javase/17/docs/api/java.base/javax/net/ssl/TrustManagerFactory.html#getTrustManagers()

File: kroxylicious-kms-provider-hashicorp-vault/src/main/java/io/kroxylicious/kms/provider/hashicorp/vault/config/InsecureTls.java (hunk 1 above)
Comment (SamBarker) [reply kroxylicious-kms-provider-hashicorp-vault/src/main/java/io/kroxylicious/kms/provider/hashicorp/vault/config/InsecureTls.java:16]: If we don't supply a secure version by default though I worry people will just run this anyway. Sonar threw its toys out of the pram about it.

I'd be quite comfortable if we had it as an optional module that required something additional to enable it. I just worry a lot about providing a proxy which messes trust/tls up in production

File: kroxylicious-kms-provider-hashicorp-vault/src/main/java/io/kroxylicious/kms/provider/hashicorp/vault/config/TrustProvider.java (hunk 2 above)
Comment (SamBarker) [reply kroxylicious-kms-provider-hashicorp-vault/src/main/java/io/kroxylicious/kms/provider/hashicorp/vault/config/TrustProvider.java:25]: Ah if it's not our api then leave it as an array. I'm on the fence about null handling. I'd rather get rid of it but equally API consistency is important

File: kroxylicious-kms-provider-hashicorp-vault/src/main/java/io/kroxylicious/kms/provider/hashicorp/vault/config/InsecureTls.java (hunk 1 above)
Comment (robobario) [reply kroxylicious-kms-provider-hashicorp-vault/src/main/java/io/kroxylicious/kms/provider/hashicorp/vault/config/InsecureTls.java]: +1 to having it available and documented as dangerous.

What would the secure version by default be for a Vault instance with private certs? They are going to have to supply a trust store and configure something.

File: kroxylicious-kms-provider-hashicorp-vault/src/test/java/io/kroxylicious/kms/provider/hashicorp/vault/VaultKmsIT.java (hunk 4 above)
Comment (robobario) [reply kroxylicious-kms-provider-hashicorp-vault/src/test/java/io/kroxylicious/kms/provider/hashicorp/vault/VaultKmsIT.java]: turns out we didn't need to set it anyway, not sure what changed since I was seeing permissions errors a few days ago. Maybe it's the refactor to bouncycastle and the key files have looser permissions than what was output by openssl and keytool.

File: kroxylicious-kms-provider-hashicorp-vault/src/main/java/io/kroxylicious/kms/provider/hashicorp/vault/config/Tls.java (hunk 5 above)
Comment (robobario) [reply kroxylicious-kms-provider-hashicorp-vault/src/main/java/io/kroxylicious/kms/provider/hashicorp/vault/config/Tls.java]: I've moved the configuration classes into kroxylicious-api. This time I've put no logic about `TrustManager`s into the config classes, the only bit of smarts is password file classes know how to load the password.

The two clients (kroxylicious-runtime and the vault KMS) use a visitor to get at the concrete TrustProvider type and then each have their own logic for constructing SSL contexts.

I also wondered about using a sealed hierarchy and switching but we need a higher lang level to support the pattern matching switch.

File: kroxylicious-kms-provider-hashicorp-vault/src/main/java/io/kroxylicious/kms/provider/hashicorp/vault/config/InsecureTls.java (hunk 1 above)
Comment (robobario) [reply kroxylicious-kms-provider-hashicorp-vault/src/main/java/io/kroxylicious/kms/provider/hashicorp/vault/config/InsecureTls.java]: I've refactored it again to share the configuration classes that we use for configuring upstream/downstream TLS in the kroxy runtime. So for this PR I think they should remain consistent.

File: kroxylicious-kms-provider-hashicorp-vault/src/test/java/io/kroxylicious/kms/provider/hashicorp/vault/VaultKmsIT.java (hunk 3 above)
Comment (robobario) [reply kroxylicious-kms-provider-hashicorp-vault/src/test/java/io/kroxylicious/kms/provider/hashicorp/vault/VaultKmsIT.java:99]: mm, false positives are possible through misconfiguration yes. If you want to exercise TLS and configure the Vault KMS to point at an `http://` endpoint then all the requests will succeed.

Maybe the testing is at the wrong level anyway, I imagine we don't need to test every KMS interaction to check that TLS is configured for each and every one. Maybe this set of tests should run only against one vault configuration and we make a second set of tests that are just for checking TLS through a single API call.

Comment (robobario) [reply kroxylicious-kms-provider-hashicorp-vault/src/test/java/io/kroxylicious/kms/provider/hashicorp/vault/VaultKmsIT.java]: pushed up, the VaultKmsIT checks that the KMS business logic and that it's fulfilling some contractual obligations and the VaultKmsTlsIT just wants to check that we can make an API call with various TLS configurations against a vault that only exposes a TLS endpoint (and a negative test showing that we cannot connect without the right configuration).

File: kroxylicious-kms-provider-hashicorp-vault/src/main/java/io/kroxylicious/kms/provider/hashicorp/vault/config/Tls.java (hunk 5 above)
Comment (robobario) [reply kroxylicious-kms-provider-hashicorp-vault/src/main/java/io/kroxylicious/kms/provider/hashicorp/vault/config/Tls.java]: if using Visitor is a bit overwrought, happy enough to switch to instanceof, what do you all prefer?

File: kroxylicious-api/src/main/java/io/kroxylicious/proxy/config/tls/FilePassword.java [hunk 7]
```diff
@@ -41,9 +45,6 @@ public String toString() {
 
     @Nullable
     static String readPasswordFile(String passwordFile) {
-        if (passwordFile == null) {
```
Comment (k-wall) [inline kroxylicious-api/src/main/java/io/kroxylicious/proxy/config/tls/FilePassword.java:44]: Remove the @Nullable return to be consistent with your change.

Comment (k-wall) [reply kroxylicious-api/src/main/java/io/kroxylicious/proxy/config/tls/FilePassword.java:44]: Also let's make the exception message at least name the file.

File: kroxylicious-api/src/test/java/io/kroxylicious/proxy/config/tls/FilePasswordTest.java [hunk 8]
```diff
@@ -62,7 +68,17 @@ void toStringDoesNotLeakPassword() throws Exception {
         Files.writeString(file.toPath(), password);
         var provider = new FilePassword(file.getAbsolutePath());
         assertThat(provider)
-                .extracting(FilePassword::toString)
+                .extracting(Object::toString)
+                .doesNotHave(new Condition<>(s -> s.contains(password), "contains password"));
+    }
+
+    @Test
+    void toStringDoesNotLeakPasswordFilePath() throws Exception {
```
Comment (k-wall) [inline kroxylicious-api/src/test/java/io/kroxylicious/proxy/config/tls/FilePasswordTest.java:76]:
```suggestion
    void filePasswordToStringDoesNotLeakPassword() throws Exception {
```

File: kroxylicious-kms-provider-hashicorp-vault/src/main/java/io/kroxylicious/kms/provider/hashicorp/vault/config/JdkTls.java [hunk 9]
```diff
@@ -0,0 +1,134 @@
...
+    @NonNull
+    private SSLContext customSSLContext(TrustProvider trust) throws NoSuchAlgorithmException, KeyManagementException {
+        Optional<TrustManager[]> trustManagers = getTrustManagers(trust);
+        return trustManagers.map(JdkTls::getSslContext).orElse(SSLContext.getDefault());
+    }
+
+    @NonNull
+    private static SSLContext getSslContext(TrustManager[] trustManagers1) {
+        try {
+            SSLContext rsa = SSLContext.getInstance("TLS");
```
Comment (k-wall) [inline kroxylicious-kms-provider-hashicorp-vault/src/main/java/io/kroxylicious/kms/provider/hashicorp/vault/config/JdkTls.java:96]: Let's respect the environment's default protocol name.   
```suggestion
            SSLContext rsa = SSLContext.getInstance(SSLContext.getDefault().getProtocol());
```

File: kroxylicious-kms-provider-hashicorp-vault/src/main/java/io/kroxylicious/kms/provider/hashicorp/vault/config/JdkTls.java [hunk 10]
```diff
@@ -0,0 +1,134 @@
...
+                }
+            }
+
+            @Override
+            public Optional<TrustManager[]> visit(InsecureTls insecureTls) {
+                if (insecureTls.insecure()) {
+                    return Optional.of(INSECURE_TRUST_MANAGERS);
+                }
+                else {
+                    return Optional.empty();
```
Comment (k-wall) [inline kroxylicious-kms-provider-hashicorp-vault/src/main/java/io/kroxylicious/kms/provider/hashicorp/vault/config/JdkTls.java:129]: Maybe just returning the platform's default trust managers would be consistent.   I think that means the need for the Optional goes..



```suggestion
                        return Optional.of(TrustManagerFactory.getInstance(TrustManagerFactory.getDefaultAlgorithm()).getTrustManagers());
```

File: kroxylicious-kms-provider-hashicorp-vault/src/test/java/io/kroxylicious/kms/provider/hashicorp/vault/VaultKmsIT.java [hunk 11]
```diff
@@ -45,30 +45,28 @@ class VaultKmsIT {
...
     @SuppressWarnings("rawtypes")
-    private VaultContainer vaultContainer;
-    private VaultKms service;
+    private static VaultContainer vaultContainer;
+    private static VaultKms service;
 
-    @BeforeEach
-    @SuppressWarnings("resource")
-    void beforeEach() {
+    @BeforeAll
```
Comment (k-wall) [inline kroxylicious-kms-provider-hashicorp-vault/src/test/java/io/kroxylicious/kms/provider/hashicorp/vault/VaultKmsIT.java:51]: What was the motivation for the lifecycle change?  I tend to use beforeEach for the certainty that there's no contamination between tests, unless the overheads are just too high.  Docker containers spin up really quickly, so in this case I went beforeEach.  

I'm not against the change, just curious about your thought process.

File: kroxylicious-kms-provider-hashicorp-vault/src/test/java/io/kroxylicious/kms/provider/hashicorp/vault/VaultKmsTlsIT.java [hunk 12]
```diff
@@ -0,0 +1,188 @@
...
+                // internal vault commands we make while execed into the container can communicate with vault over plain
+                // without having to trust the TLS cert within the container. We do not expose the plain listener to
+                // prevent false positives if the test connected to the wrong endpoint.
+                .withExposedPorts(TLS_PORT)
+                .withCopyFileToContainer(MountableFile.forClasspathResource("Vault.hcl"), "/vault/config/Vault.hcl")
+                .withCopyFileToContainer(MountableFile.forHostPath(keys.serverCertPem()), "/vault/config/cert.pem")
+                .withCopyFileToContainer(MountableFile.forHostPath(keys.serverPrivateKeyPem()), "/vault/config/key.pem")
+                .withInitCommand(
+                        "secrets enable transit");
+        vaultContainer.setWaitStrategy(Wait.forHttp("/v1/sys/health").forPort(TLS_PORT).usingTls().allowInsecure().forStatusCode(200));
```
Comment (k-wall) [inline kroxylicious-kms-provider-hashicorp-vault/src/test/java/io/kroxylicious/kms/provider/hashicorp/vault/VaultKmsTlsIT.java:95]: you add a wait strategy here, but not to the non-TLS test case.  Is that intentional?

File: kroxylicious-kms-provider-hashicorp-vault/src/test/java/io/kroxylicious/kms/provider/hashicorp/vault/VaultKmsTlsIT.java [hunk 13]
```diff
@@ -0,0 +1,188 @@
...
+    }
+
+    @AfterAll
+    public static void afterAll() {
+        vaultContainer.stop();
+    }
+
+    @Test
+    void tlsConnectionFailsWithoutClientTrust() {
+        // passing null tls means we do not attempt to obtain a custom SSLContext from FilterFactoryContext
```
Comment (k-wall) [inline kroxylicious-kms-provider-hashicorp-vault/src/test/java/io/kroxylicious/kms/provider/hashicorp/vault/VaultKmsTlsIT.java:106]:
```suggestion
        // passing null tls means we do not attempt to obtain a custom SSLContext
```

File: kroxylicious-kms-provider-hashicorp-vault/src/test/java/io/kroxylicious/kms/provider/hashicorp/vault/VaultKmsTlsIT.java [hunk 14]
```diff
@@ -0,0 +1,188 @@
...
+                .succeedsWithin(Duration.ofSeconds(5))
+                .isEqualTo(keyName);
+    }
+
+    private ReadKeyData createKek(String keyId) {
+        return runVaultCommand(new TypeReference<>() {
+        }, "vault", "write", "-f", "transit/keys/%s".formatted(keyId));
+    }
+
+    private <D> D runVaultCommand(TypeReference<VaultResponse<D>> valueTypeRef, String... args) {
```
Comment (k-wall) [inline kroxylicious-kms-provider-hashicorp-vault/src/test/java/io/kroxylicious/kms/provider/hashicorp/vault/VaultKmsTlsIT.java:136]: we should extract the common methods.

File: kroxylicious-kms-provider-hashicorp-vault/src/test/java/io/kroxylicious/kms/provider/hashicorp/vault/VaultKmsTlsIT.java [hunk 15]
```diff
@@ -0,0 +1,188 @@
...
+        vaultKmsMap.put("jksTls", getTlsVaultKms(tlsForTrustStoreInlinePassword(keys.jksClientTruststore())));
+        vaultKmsMap.put("defaultStoreTypeTls", getTlsVaultKms(defaultStoreTypeTls(keys.jksClientTruststore())));
+        vaultKmsMap.put("tlsInsecure", getTlsVaultKms(insecureTls()));
+    }
+
+    private static void startVault() {
+        vaultContainer = new VaultContainer<>(HASHICORP_VAULT)
+                .withVaultToken(VAULT_TOKEN)
+                .withEnv("VAULT_FORMAT", "json")
+                // in vault server -dev mode the 8200 port cannot be reconfigured to TLS and is also useful so that the
```
Comment (k-wall) [inline kroxylicious-kms-provider-hashicorp-vault/src/test/java/io/kroxylicious/kms/provider/hashicorp/vault/VaultKmsTlsIT.java:85]: the comment is quite hard to parse.

File: kroxylicious-kms-provider-hashicorp-vault/src/test/java/io/kroxylicious/kms/provider/hashicorp/vault/config/InsecureTlsTest.java [hunk 16]
```diff
@@ -0,0 +1,58 @@
...
+        JdkTls insecureTls = new JdkTls(new Tls(null, new InsecureTls(false)));
+        assertThat(insecureTls.sslContext()).isSameAs(SSLContext.getDefault());
+    }
+
+    @Test
+    void testInsecureTlsEnabled() {
+        InsecureTls insecureTls = new InsecureTls(true);
+        Optional<TrustManager[]> maybeTrustManagers = JdkTls.getTrustManagers(insecureTls);
+        assertThat(maybeTrustManagers).isNotEmpty().hasValueSatisfying(trustManagers -> {
+            assertThat(trustManagers).hasSize(1).containsExactly(INSECURE_TRUST_MANAGER);
```
Comment (k-wall) [inline kroxylicious-kms-provider-hashicorp-vault/src/test/java/io/kroxylicious/kms/provider/hashicorp/vault/config/InsecureTlsTest.java:38]: non-blocking.. the fact that it is using *the* INSECURE_TRUST_MANAGER is really an implementational concern.  The unit test shouldn't care.  A better way would be to create a certificate and offer it to the trust managers and make sure they accept it.

**Review by k-wall — APPROVED** [review]

**Comment by k-wall** [comment]
Couple of niggles, but on the whole the change looks good to me.  Thank you @robobario

File: kroxylicious-kms-provider-hashicorp-vault/src/test/java/io/kroxylicious/kms/provider/hashicorp/vault/VaultKmsIT.java (hunk 11 above)
Comment (robobario) [reply kroxylicious-kms-provider-hashicorp-vault/src/test/java/io/kroxylicious/kms/provider/hashicorp/vault/VaultKmsIT.java:51]: The tests take ~2seconds each for me, earlier I had the branch running the 7 existing tests against all the TLS configurations so it was taking quite a long time. I didn't want to add another minute to the suite. But since refactoring to have a smaller TLS suite I think it's better to go back to the cleaner way.

In general I worry that the IT suite is steadily pushing the build time up and that the time to hear back from sonar is getting quite big. Maybe we should at making it parallelisable so we can have 2 surefire forks.

Comment (SamBarker) [reply kroxylicious-kms-provider-hashicorp-vault/src/test/java/io/kroxylicious/kms/provider/hashicorp/vault/VaultKmsIT.java:51]: I'm not sure there is much to win from parallelisation in the Git Hub action runners as they are iirc 2core. 

I'm tempted to say we change the trigger on the sonar run so it doesn't wait for a successful build. Or the more drastic open we restructure the build pipeline so that we have what I would term a commit build run which then triggers sonar and the integration tests in parallel. By a commit build I mean compile and unit test run, plus associated formatting checks/fixes.

File: kroxylicious-kms-provider-hashicorp-vault/src/test/java/io/kroxylicious/kms/provider/hashicorp/vault/VaultKmsTlsIT.java (hunk 12 above)
Comment (robobario) [reply kroxylicious-kms-provider-hashicorp-vault/src/test/java/io/kroxylicious/kms/provider/hashicorp/vault/VaultKmsTlsIT.java:95]: Yes, the default needs to be overridden for our TLS usage. In the TLS test we only expose the TLS port (to prevent any chance of accidentally connecting to the plain port) and the testcontainers wait code fails to connect because it doesn't trust the cert.

File: kroxylicious-api/src/test/java/io/kroxylicious/proxy/config/tls/FilePasswordTest.java (hunk 8 above)
Comment (robobario) [reply kroxylicious-api/src/test/java/io/kroxylicious/proxy/config/tls/FilePasswordTest.java]: changed to parameterized tests

File: kroxylicious-kms-provider-hashicorp-vault/src/main/java/io/kroxylicious/kms/provider/hashicorp/vault/config/JdkTls.java (hunk 10 above)
Comment (robobario) [reply kroxylicious-kms-provider-hashicorp-vault/src/main/java/io/kroxylicious/kms/provider/hashicorp/vault/config/JdkTls.java]: have updated to do this, you also have to init the TrustManagerFactory with a `null` KeyContext which eventually returns a TrustManager that trusts default certs like `trustManager = getInstance(TrustStoreManager.getTrustedCerts());`. It's not really documented that that should be the behaviour for null though. We have IT and unit tests showing it rejects the server self-signed-cert, but can we depend on this across platforms?

```
    /**
     * Initializes this factory with a source of certificate
     * authorities and related trust material.
     * <P>
     * The provider typically uses a KeyStore as a basis for making
     * trust decisions.
     * <P>
     * For more flexible initialization, please see
     * {@link #init(ManagerFactoryParameters)}.
     *
     * @param ks the key store, or null
     * @throws KeyStoreException if this operation fails
     */
    public final void init(KeyStore ks) throws KeyStoreException {
        factorySpi.engineInit(ks);
    }
    ```

File: kroxylicious-kms-provider-hashicorp-vault/src/test/java/io/kroxylicious/kms/provider/hashicorp/vault/VaultKmsTlsIT.java (hunk 14 above)
Comment (robobario) [reply kroxylicious-kms-provider-hashicorp-vault/src/test/java/io/kroxylicious/kms/provider/hashicorp/vault/VaultKmsTlsIT.java]: Created a TestVault class that hides the differences between TLS/non-TLS, unifies constructing the container and offers `runVaultCommand` along with methods to obtain the token and endpoint.

File: kroxylicious-kms-provider-hashicorp-vault/src/test/java/io/kroxylicious/kms/provider/hashicorp/vault/VaultKmsIT.java (hunk 11 above)
Comment (robobario) [reply kroxylicious-kms-provider-hashicorp-vault/src/test/java/io/kroxylicious/kms/provider/hashicorp/vault/VaultKmsIT.java]: Have refactored both ITs back to creating a vault per test. The vault kms tests take ~30seconds so something to keep an eye on.

File: kroxylicious-kms-provider-hashicorp-vault/src/test/java/io/kroxylicious/kms/provider/hashicorp/vault/config/InsecureTlsTest.java (hunk 16 above)
Comment (robobario) [reply kroxylicious-kms-provider-hashicorp-vault/src/test/java/io/kroxylicious/kms/provider/hashicorp/vault/config/InsecureTlsTest.java]: Have updated the tests to check that the TrustManagers reject or accept a self-signed-cert

File: kroxylicious-api/src/test/java/io/kroxylicious/proxy/config/tls/TlsTest.java [hunk 17]
```diff
@@ -0,0 +1,39 @@
...
+import static org.assertj.core.api.Assertions.assertThat;
+
+class TlsTest {
+
+    @Test
+    void testGetStoreTypeOrPlatformDefault() {
+        assertThat(Tls.getStoreTypeOrPlatformDefault("PEM")).isEqualTo("PEM");
+        assertThat(Tls.getStoreTypeOrPlatformDefault("JKS")).isEqualTo("JKS");
+        assertThat(Tls.getStoreTypeOrPlatformDefault("jKs")).isEqualTo("JKS");
+        assertThat(Tls.getStoreTypeOrPlatformDefault("jks")).isEqualTo("JKS");
```
Comment (SamBarker) [inline kroxylicious-api/src/test/java/io/kroxylicious/proxy/config/tls/TlsTest.java:23]: should we be testing other capitalisations of `PEM`?

File: kroxylicious-kms-provider-hashicorp-vault/src/test/java/io/kroxylicious/kms/provider/hashicorp/vault/VaultKmsTlsIT.java [hunk 18]
```diff
@@ -0,0 +1,142 @@
...
+class VaultKmsTlsIT {
+
+    private static final String VAULT_TOKEN = "token";
+    private TestVault vaultContainer;
+
+    private static final CertificateGenerator.Keys keys = CertificateGenerator.generate();
+
+    @BeforeEach
+    void beforeEach() {
+        assumeThat(DockerClientFactory.instance().isDockerAvailable()).withFailMessage("docker unavailable").isTrue();
```
Comment (SamBarker) [inline kroxylicious-kms-provider-hashicorp-vault/src/test/java/io/kroxylicious/kms/provider/hashicorp/vault/VaultKmsTlsIT.java:51]: I wonder if there is a better way to do this. I have a feeling there is some sort of `DisabledIf` type annotation. Anyway not for this PR.

**Review by SamBarker — APPROVED** [review]
LGTM

**Comment by robobario** [comment]
Just realiased one last inconsistency is that netty handles PEM parsing for us in kroxy core. We would need further work to support PEM format for vault.

File: kroxylicious-api/src/test/java/io/kroxylicious/proxy/config/tls/TlsTest.java (hunk 17 above)
Comment (robobario) [reply kroxylicious-api/src/test/java/io/kroxylicious/proxy/config/tls/TlsTest.java:23]: the input is arbitrary as it uppercases whatever you give it.

Comment (SamBarker) [reply kroxylicious-api/src/test/java/io/kroxylicious/proxy/config/tls/TlsTest.java:23]: cool it just seemed strange to see a mix of cases for one name and not the other...
